File: gaussian.py
import matplotlib.pyplot as plt
import numpy as np
from sklearn.gaussian_process import GaussianProcessRegressor
from sklearn.gaussian_process.kernels import RBF, ConstantKernel as C
import os
import sys
import warnings
import time
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error
from sklearn.gaussian_process.kernels import Matern
from sklearn.gaussian_process.kernels import ExpSineSquared
from sklearn.gaussian_process.kernels import DotProduct, WhiteKernel,ConstantKernel
from data_prepare.data_load import data_loader
from data_prepare.accuracy import accuracy
import glob
warnings.filterwarnings("ignore")

###Hyperparameters
noise_filter = False #Must be set to False!!!
png_name = "const_RBF_noise" # Set the name of the output PNG file. PNG files will be saved in the "result" folder.

# Do not change the following two parameters.
calculate_time = True # Calculate the time of calculation especially about GPR
test_set_ratio = 0.8 # Must be set to 0.8!!!

### Kernel Setting
const = ConstantKernel()
rbf = RBF(0.1, (1e-2, 1e1))
matern = Matern(length_scale=1.0, nu=1.5)
exp_sine = ExpSineSquared(length_scale=1.0, periodicity=5.0, periodicity_bounds=(1, 10))
dot_product = DotProduct(sigma_0=1.0, sigma_0_bounds=(1e-3, 1e3))
white = WhiteKernel()
kernel_set = [const, rbf, matern, exp_sine, dot_product, white]

#### Kernel Model Set
kernel_model = [ConstantKernel(1.0, (1e-3, 1e3)) * DotProduct() + WhiteKernel(),
           ConstantKernel(1.0, (1e-3, 1e3)) * RBF(0.1, (1e-2, 1e1)) + WhiteKernel(),
           ConstantKernel(1.0, (1e-3, 1e3)) * RBF(0.1, (1e-2, 1e1)) + WhiteKernel() + ConstantKernel() * DotProduct(),
           ConstantKernel(1.0, (1e-3, 1e3)) * Matern(nu=1.5) + WhiteKernel(),
           ConstantKernel(1.0, (1e-3, 1e3)) * Matern(nu=1.5) + WhiteKernel() + ConstantKernel() * DotProduct(),
           ConstantKernel(1.0, (1e-3, 1e3)) * Matern(nu=0.5) + WhiteKernel(),
           ConstantKernel(1.0, (1e-3, 1e3)) * Matern(nu=0.5) + WhiteKernel() + ConstantKernel() * DotProduct(),
           ConstantKernel(1.0, (1e-3, 1e3)) * Matern(nu=2.5) + WhiteKernel(),
           ConstantKernel(1.0, (1e-3, 1e3)) * Matern(nu=2.5) + WhiteKernel() + ConstantKernel() * DotProduct()]
#from https://github.com/hkaneko1985/dcekit/blob/master/demo_gp_kernel_design_test.py 


# kernel = kernel_model[0]
kernel = const * rbf 

#### RBF Setting
# See https://scikit-learn.org/stable/modules/generated/sklearn.gaussian_process.GaussianProcessRegressor.html for further information
# kernel = default ConstantKernel(1.0, constant_value_bounds="fixed") * RBF(1.0, length_scale_bounds="fixed")
# alpha : (maybe) adding noise on the training data
# n_restarts_optimizer: 対数周辺尤度を最大化するための最適化を，異なる初期値から何回再起動するかを設定
gpr = GaussianProcessRegressor(kernel=kernel, n_restarts_optimizer=50, alpha=0.1)
# 1) Load data
# Find all .dat files in data_prepare directory

dat_files = glob.glob('data_prepare/*.dat')
print(f"Found {len(dat_files)} .dat files")

# Load and combine all .dat files
data_list = []
for file in dat_files:
  try:
    file_data = np.loadtxt(file, delimiter=',')
    # Check and adjust the orientation if needed
    if file_data.ndim == 2 and file_data.shape[0] == 3:
        # Data is already in the correct format (3 rows)
        data_list.append(file_data)
    elif file_data.ndim == 2 and file_data.shape[1] == 3:
        # Data has 3 columns, transpose to get 3 rows
        data_list.append(file_data.T)
    else:
        print(f"Skipping {file}: incompatible shape {file_data.shape}")
        continue
    print(f"Loaded {file}: shape after processing {data_list[-1].shape}")
  except Exception as e:
    print(f"Error loading {file}: {e}")

# Make sure we have data to process
if not data_list:
    print("No valid data files could be loaded!")
    sys.exit(1)
def hampel_filter(vals, window_size=7, n_sigmas=3):
  """
  vals: 1D numpy array
  window_size: 偶数可、窓幅
  n_sigmas: 外れ値とみなす閾値
  """
  vals = vals.copy()
  L = len(vals)
  k = window_size // 2
  for i in range(L):
    start = max(0, i - k)
    end   = min(L, i + k + 1)
    window = vals[start:end]
    med = np.median(window)
    mad = 1.4826 * np.median(np.abs(window - med))
    if mad > 0 and np.abs(vals[i] - med) > n_sigmas * mad:
      vals[i] = med
  return vals
# Concatenate horizontally (along columns) since data is in row format
data = np.hstack(data_list)
print(f"Combined data shape: {data.shape}")
omega_raw, sys_gain_raw, arg_g_raw = data
SysGain_f = hampel_filter(sys_gain_raw, window_size=7, n_sigmas=3)
argG_f    = hampel_filter(arg_g_raw,    window_size=7, n_sigmas=3)

# 2) Sort data by frequency
idx = np.argsort(omega_raw)
omega = omega_raw[idx]
sys_gain = SysGain_f[idx]
arg_g = argG_f  [idx]
# → Hampel フィルターをかける
# 3) Prepare input X and output y
X = omega.reshape(-1, 1)
y_gain = np.log10(sys_gain)
y_phase = arg_g

# 4) Setup GPR for gain (using the kernel already defined above)
gpr_gain = GaussianProcessRegressor(kernel=kernel, n_restarts_optimizer=9, alpha=0.1)
gpr_gain.fit(X, y_gain)

# 5) Setup GPR for phase
# gpr_phase = GaussianProcessRegressor(kernel=kernel, n_restarts_optimizer=9, alpha=0.1)
# gpr_phase.fit(X, y_phase)

# 6) Create prediction frequency grid
omega_test = np.logspace(np.log10(omega.min()), np.log10(omega.max()), 500)
X_test = omega_test.reshape(-1, 1)

# 7) Make predictions
y_gain_pred, y_gain_std = gpr_gain.predict(X_test, return_std=True)
# y_phase_pred, y_phase_std = gpr_phase.predict(X_test, return_std=True)

# 8) Plot gain
plt.figure(figsize=(10, 6))
plt.semilogx(omega, y_gain, 'b*', markersize=6, label='Observed (gain)')
plt.semilogx(omega_test, y_gain_pred, 'r-', linewidth=2, label='GPR fit')
plt.fill_between(omega_test,
        y_gain_pred - 2*y_gain_std,
        y_gain_pred + 2*y_gain_std,
        alpha=0.2, color='r')
plt.xlabel('ω [rad/s]', fontsize=16)
plt.ylabel('20 log₁₀|G(jω)| [dB]', fontsize=16)
plt.title('Bode Gain Plot with GPR', fontsize=16)
plt.legend(fontsize=12, loc='best')
plt.grid(True)
plt.ylim(-50,10)
plt.savefig(f"result/{png_name}_gain.png")
plt.close()

# # 9) Plot phase
plt.figure(figsize=(10, 6))
plt.semilogx(omega, y_phase, 'b*', markersize=6, label='Observed (phase)')
plt.semilogx(omega_test, y_phase_pred, 'r-', linewidth=2, label='GPR fit')
plt.fill_between(omega_test,
        y_phase_pred - 2*y_phase_std,
        y_phase_pred + 2*y_phase_std,
        alpha=0.2, color='r')
plt.xlabel('ω [rad/s]', fontsize=16)
plt.ylabel('Phase [rad]', fontsize=16)
plt.title('Bode Phase Plot with GPR', fontsize=16)
plt.legend(fontsize=12, loc='best')
plt.grid(True)
plt.savefig(f"result/{png_name}_phase.png")
plt.close()

# 10) Nyquist plot
G_dataset = sys_gain * np.exp(1j * arg_g)
H_best = 10**(y_gain_pred/20) * np.exp(1j * y_phase_pred)

plt.figure(figsize=(10, 6))
plt.plot(G_dataset.real, G_dataset.imag, 'b*', markersize=6, label='Data')
plt.plot(H_best.real, H_best.imag, 'r-', linewidth=2, label='GPR Est.')
plt.xlabel('Re', fontsize=16)
plt.ylabel('Im', fontsize=16)
plt.title('Nyquist Plot', fontsize=16)
plt.grid(True)
plt.legend(fontsize=12)
plt.tight_layout()
plt.savefig(f"result/{png_name}_nyquist.png")
plt.close()

File: gaussian.m

clear; clc;

%% ---------------- ユーザ設定 ------------------------
noiseFilter    = false;               % true: Hampel で外れ値除去 (falseに変更してテスト、必要ならtrueに戻す)
pngName        = "const_RBF_white";   % 出力 PNG (result フォルダに保存)
calculateTime  = true;                % 計算時間を表示
testSetRatio   = 0.8;                 % テストデータの割合 (訓練データは 1-testSetRatio = 0.2)

% GPR カーネル : 係数付き RBF(=SquaredExponential) + WhiteNoise
kernelFcn      = 'squaredexponential';   % RBF と同等
sigma0         = 0.1;                    % 事前ノイズの初期値
numRestart     = 9;                      % ハイパーパラメータ最適化の再起動回数
%% ----------------------------------------------------

% --- データ読み込み ---
% 'result/merged.dat'ファイルが存在することを確認してください
dataPath = fullfile('result', 'merged.dat');
if ~exist(dataPath, 'file')
    error('データファイルが見つかりません: %s', dataPath);
end
tmp = importdata(dataPath, ',');

% データが期待通りに読み込まれているか確認 (例: 3行N列)
if size(tmp, 1) ~= 3
    error('データファイルの形式が不正です。3行 (omega, gain, phase) である必要があります。');
end

% --- データ整理 ---
% 周波数でソート
[omega, idx] = sort(tmp(1,:));
SysGain_raw = tmp(2,idx);
argG_raw    = tmp(3,idx);

% データの要素数 (サンプル数) を表示
n = length(omega);
disp(['データのサンプル数: ', num2str(n)]);

% --- ノイズ除去 (オプション) ---
if noiseFilter
    % Signal Processing Toolbox の hampel を試す
    if license('test', 'Signal_Toolbox') && exist('hampel', 'file')
        disp('Hampel フィルタ (Signal Processing Toolbox) を使用します。');
        SysGain = hampel(SysGain_raw, 15); % Hampelフィルタ適用 (ゲイン)
        argG    = hampel(argG_raw, 15);    % Hampelフィルタ適用 (位相)
    else
        disp('Hampel フィルタ (自家製) を使用します。');
        % 自家製 Hampel フィルタを使用する場合
        SysGain = hampelCustom(SysGain_raw, 15);
        argG    = hampelCustom(argG_raw, 15);
    end
else
    disp('Hampel フィルタは適用されません。');
    SysGain = SysGain_raw;
    argG    = argG_raw;
end
% 複素ゲイン (プロット等では使用しないが参考のため)
% G = SysGain .* exp(1i*argG); % 使わないのでコメントアウト

%% 学習データ準備
% --- 特徴量と目標値の作成 ---
% fitrgp は列ベクトルを期待するため、ここで転置 (.').
X = log10(omega).';           % 特徴量 : log10(ω) -> N x 1 列ベクトル
Y = 20*log10(SysGain).';      % 目標値 : 20*log10|G| -> N x 1 列ベクトル

% X と Y のサイズ確認 (デバッグ用)
disp(['Xのサイズ: ', num2str(size(X))]); % -> N x 1 になるはず
disp(['Yのサイズ: ', num2str(size(Y))]); % -> N x 1 になるはず

if size(X, 1) ~= size(Y, 1)
    error('内部エラー: X と Y の行数 (サンプル数) が一致しません。データ処理を確認してください。');
end
if size(X,1) ~= n || size(Y,1) ~= n
    error('内部エラー: X, Y のサンプル数が元のデータ数と一致しません。');
end


% --- 訓練 / テスト分割 (80‑20) ---
% testSetRatio はテストデータの割合
nTrain = floor(n * (1-testSetRatio)); % 訓練データの数 (例: 0.2 * n)
nTest  = n - nTrain;                  % テストデータの数 (例: 0.8 * n)

% ランダムなインデックスを生成して分割
idx = randperm(n);
trainIdx = idx(1:nTrain);
testIdx  = idx(nTrain+1:end);

% 列ベクトルとして分割
XTrain = X(trainIdx, :); % -> nTrain x 1
YTrain = Y(trainIdx, :); % -> nTrain x 1
XTest  = X(testIdx, :);  % -> nTest x 1
YTest  = Y(testIdx, :);  % -> nTest x 1

% 分割後のサイズ確認 (デバッグ用)
disp(['訓練データ数: ', num2str(nTrain)]);
disp(['テストデータ数: ', num2str(nTest)]);
disp(['XTrainのサイズ: ', num2str(size(XTrain))]);
disp(['YTrainのサイズ: ', num2str(size(YTrain))]);
disp(['XTestのサイズ: ', num2str(size(XTest))]);
disp(['YTestのサイズ: ', num2str(size(YTest))]);

if isempty(XTrain) || isempty(YTrain)
    error('訓練データが空です。データ数または testSetRatio を確認してください。');
end

%% GPR モデル学習
if calculateTime, startT = tic; end

% --- ハイパーパラメータ最適化あり ---
% ★★★ 修正点: fitrgp には訓練データ (XTrain, YTrain) を渡す ★★★
gprMdl = fitrgp( ...
    XTrain, YTrain, ... % ここを修正！ X, Y ではなく XTrain, YTrain を使う
    'KernelFunction',      kernelFcn,   ...
    'Sigma',               sigma0,      ...
    'Standardize',         true,        ...
    'OptimizeHyperparameters', 'all', ...
    'HyperparameterOptimizationOptions', ...
        struct(...
            'MaxObjectiveEvaluations', numRestart, ...
            'ShowPlots', false, ...
            'Verbose',   0) ...
);


if calculateTime
    elapsedT = toc(startT);
    fprintf("Elapsed time : %.2f s\n", elapsedT);
end

%% 予測 (細かい周波数グリッド & テスト / 訓練)
% --- 細かいグリッドでの予測 ---
% logspace はデフォルトで列ベクトルを生成する
omegaFine = logspace(log10(min(omega)), log10(max(omega)), 500).';
XFine = log10(omegaFine); % predict に渡すため、log10 を適用 (すでに列ベクトル)

% 予測実行
[YPredFine, YSD] = predict(gprMdl, XFine); % XFine は N_fine x 1

% --- 訓練/テストデータでの予測 ---
YPredTrain = predict(gprMdl, XTrain); % XTrain は nTrain x 1
YPredTest  = predict(gprMdl, XTest);  % XTest は nTest x 1

% MSE 計算
mseTrain = mean((YTrain - YPredTrain).^2);
mseTest  = mean((YTest  - YPredTest ).^2);
fprintf("Training MSE : %.4f\nTest MSE     : %.4f\n", mseTrain, mseTest);

%% プロット 1 : Bode ゲイン + GPR
figure('Units','pixels','Position',[100 100 800 520]);
hold on; grid on; box on;

% 元データ (フィルタ前)
semilogx(omega, 20*log10(SysGain_raw), 'b.', 'MarkerSize',4, ...
    'DisplayName','Raw data'); % HandleVisibility を変更して凡例に表示

% 訓練 / テスト (フィルタ後のデータ)
% YTrain, YTest は既に 20*log10(SysGain) 相当なのでそのまま使用
semilogx(10.^XTrain, YTrain, 'ro', 'MarkerSize',6, ...
    'DisplayName','Training data');
semilogx(10.^XTest , YTest , 'ms', 'MarkerSize',6, ...
    'DisplayName','Test data');

% GPR 予測平均
semilogx(omegaFine, YPredFine, 'g-', 'LineWidth',2, ...
    'DisplayName','GPR prediction');

% 信頼区間 (±2σ)
semilogx(omegaFine, YPredFine + 2*YSD, 'g--', 'HandleVisibility','off');
semilogx(omegaFine, YPredFine - 2*YSD, 'g--', 'HandleVisibility','off');
fill([omegaFine; flipud(omegaFine)], [YPredFine - 2*YSD; flipud(YPredFine + 2*YSD)], ...
     'g', 'FaceAlpha', 0.1, 'EdgeColor', 'none', 'DisplayName', '\pm 2\sigma Confidence Interval');